Simplify and de-layer Service REST implementation

thockin opened the pull request:
Simplify Service's REST implementation by using the Begin* hooks introduced in #96393.  This is a large change overall, so the PR is a series of commits that are each designed to be easy to read and review, even if they are incomplete on their own.

Background:
 
Service has an "outer" and "inner" REST handler.  This is because of how we do IP and port allocations synchronously, but since we don't have API transactions, we need to roll those back in case of a failure.  Both layers use the same `Strategy`, but the outer calls into the inner, which causes a lot of complexity in the code (including an open-coded partial reimplementation of a date-unknown snapshot of the generic REST code) and results in `Prepare` and `Validate` hooks being called twice.

This PR aims to merge those layers and use mostly generic REST code.  This DOES NOT attempt to change or fix the way IPs and ports are allocated in the Service REST stack. 

One net result of this is that the validation code *can* get a bit stricter, since it will NEVER see a service before allocation.  This is left for later.

For posterity:

The "normal" REST flow seems to be:

```
mutating webhooks
generic REST store Create {
    cleanup = BeginCreate
    BeforeCreate {
        strategy.PrepareForCreate {
            dropDisabledFields
        }
        strategy.Validate
        strategy.Canonicalize
    }
    createValidation (validating webhooks)
    storage Create
    cleanup
    AfterCreate
    Decorator
}
```

Service (before this PR) does:

```
mutating webhooks
svc custom Create {
    BeforeCreate {
        strategy.PrepareForCreate {
            dropDisabledFields
        }
        strategy.Validate
        strategy.Canonicalize
    }
    Allocations
    inner (generic) Create {
        cleanup = BeginCreate
        BeforeCreate {
            strategy.PrepareForCreate {
                dropDisabledFields
            }
            strategy.Validate
            strategy.Canonicalize
        }
        createValidation (validating webhooks)
        storage Create
        cleanup
        AfterCreate
        Decorator
    }
}
```

After this PR:

```
mutating webhooks
generic REST store Create {
    cleanup = BeginCreate
        Allocations
    BeforeCreate {
        strategy.PrepareForCreate {
            dropDisabledFields
        }
        strategy.Validate
        strategy.Canonicalize
    }
    createValidation (validating webhooks)
    storage Create
    cleanup
    AfterCreate
        Rollback allocations on error
    Decorator
}
```

/kind bug
/kind cleanup

Fixes #87603
Fixes #98994

```release-note
The `Service.spec.ipFamilyPolicy` field is now *required* in order to create or update a Service as dual-stack.  This is a breaking change from the beta behavior.  Previously the server would try to infer the value of that field from either `ipFamilies` or `clusterIPs`, but that caused ambiguity on updates.  Users who want a dual-stack Service MUST specify `ipFamilyPolicy` as either "PreferDualStack" or "RequireDualStack".  
```

cc @khenidak @aojea
FYI @liggitt @deads2k @lavalamp


For followup PR(s):
- Validation overhaul (TODO comments "validation should not pass with empty clusterIP")
  - lots of things verified twice in update path
  - lots of things not verified at all
  - hard to follow
  - tests should state which fields expect errors
  - tests should have a "base" object + list of mutations
- Audit strategy code and tests
  - use svctest
- Test perf (move test boilerplate to outer loops?)
- rename IPFamilyPolicyType and InternalTraffic. - remove Type suffix
- rename ServiceExternalTrafficPolicyTypeLocal -> ServiceTrafficPolicyLocal ?
- rename api.IPv4Protocol
- Logging cleanup and internal Error for IP failures
- make DryRun work the same in NodePort as in IP allocator?
- make dryRun return valid, in-range values ?
- TODO in ResourceLocation()
- Other TODOs
- Scrub allocator code - refactor, renames, comments.
- testcase for dryrun for IPs, NPs, and HCNP (HCNP missing now).  Lean more on {type=LB, ETP=Local} as a case.
- tests for create-on-update
- more feature tests

thockin committed:
Svc REST: Add stub begin* hooks

These will be used in the next set of commits to de-0layer service REST.

thockin committed:
Svc REST: De-layer Create

Gut the "outer" Create() and move it to the inner BeginCreate().  This
uses a "transaction" type to make cleanup functions easy to read.

Background:

Service has an "outer" and "inner" REST handler.  This is because of how we do IP and port allocations synchronously, but since we don't have API transactions, we need to roll those back in case of a failure.  Both layers use the same `Strategy`, but the outer calls into the inner, which causes a lot of complexity in the code (including an open-coded partial reimplementation of a date-unknown snapshot of the generic REST code) and results in `Prepare` and `Validate` hooks being called twice.

The "normal" REST flow seems to be:

```
mutating webhooks
generic REST store Create {
    cleanup = BeginCreate
    BeforeCreate {
        strategy.PrepareForCreate {
            dropDisabledFields
        }
        strategy.Validate
        strategy.Canonicalize
    }
    createValidation (validating webhooks)
    storage Create
    cleanup
    AfterCreate
    Decorator
}
```

Service (before this commit) does:

```
mutating webhooks
svc custom Create {
    BeforeCreate {
        strategy.PrepareForCreate {
            dropDisabledFields
        }
        strategy.Validate
        strategy.Canonicalize
    }
    Allocations
    inner (generic) Create {
        cleanup = BeginCreate
        BeforeCreate {
            strategy.PrepareForCreate {
                dropDisabledFields
            }
            strategy.Validate
            strategy.Canonicalize
        }
        createValidation (validating webhooks)
        storage Create
        cleanup
        AfterCreate
        Decorator
    }
}
```

After this commit:

```
mutating webhooks
generic REST store Create {
    cleanup = BeginCreate
        Allocations
    BeforeCreate {
        strategy.PrepareForCreate {
            dropDisabledFields
        }
        strategy.Validate
        strategy.Canonicalize
    }
    createValidation (validating webhooks)
    storage Create
    cleanup
    AfterCreate
        Rollback allocations on error
    Decorator
}
```

This same fix pattern will be applied to Delete and Update in subsequent
commits.

thockin committed:
Svc REST: Encapsulate IP and Port allocator logic

Encapsulate the allocator logic so it can be shared across REST
layers while we stage a series of commits to get rid of one layer.

thockin committed:
Svc REST: move allocator methods -> alloc object

Move all allocator-related methods onto the alloc object so it can be
used in either REST layer.  There's an INORDINATE amount of test code
here and I am skeptical that it is all useful.  That's for later
commits.

thockin committed:
Svc REST: Don't call validation directly

The validation is called soon after anyway.

harche commented:
/test pull-kubernetes-node-crio-e2e

khenidak commented on pkg/registry/core/service/storage/transaction.go:
```diff
@@ -0,0 +1,62 @@
+/*
```
This should be generalized. I am sure we have we have objects out there that have similar semantics. even if we don't we can have that now.. we will eventually will.

thockin replied:
yeah, I'm following the "start local" mindset.  If we have others that are similar we can move this up and out LATER

khenidak commented on pkg/registry/core/service/storage/transaction.go:
```diff
@@ -0,0 +1,62 @@
+
+// transaction represents something that may need to be finalized on success or
+// failure of the larger transaction.
+type transaction interface {
```
it is an operation really. not transaction. I fear that people will get the wrong ideas about the idempotency/acid properties of "a transaction". e.g. COMMIT may not work completely.

thockin replied:
"operation" doesn't capture the semantics I want.  "pending"?  "committable"?  I don't care much, I thought txn was close enough

khenidak commented on pkg/registry/core/service/storage/transaction.go:
```diff
@@ -0,0 +1,62 @@
+type transaction interface {
+	// Commit tells the transaction to finalize any changes it may have
+	// pending.
+	Commit()
```
So how would commit/revert report errors?

thockin replied:
They don't.  I have no idea what to do if your commit fails - the REST operation is already done.  So I make it your problem.

khenidak replied:
so we expect to do 

t := transaction{
commit : func(){
// handle error here?
}
}

khenidak commented on pkg/registry/core/service/ipallocator/allocator.go:
```diff
@@ -85,13 +88,16 @@ func NewAllocatorCIDRRange(cidr *net.IPNet, allocatorFactory allocator.Allocator
 		}
 	} else {
+		family = api.IPv4Protocol
 		// Don't use the IPv4 network's broadcast address.
```
low pri. We should have a note here that we depend mask size (in config ref: https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/options/validation.go#L89) validation. if validation is removed then we need to max it to `max(unsiged short)`

aojea replied:
that is done by utilnet.RangeSize in L88 https://github.com/kubernetes/utils/blob/4a2b36d8d73f0bfa1092e740605b66db34fca1cd/net/net.go#L192-L204

fedebongio commented:
/assign @caesarxuchao 
would you mind taking a look? 
/triage accepted

thockin commented:
@fedebongio @caesarxuchao  no need for anyone to look at this yet except morbid curiosity :)

khenidak commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -146,14 +146,25 @@ type Store struct {
 	// CreateStrategy implements resource-specific behavior during creation.
 	CreateStrategy rest.RESTCreateStrategy
+	// BeginCreate is an optional hook that returns a "transaction-like"
+	// commit/cleanup function which will be called at the end of operation,
```
nit
s/commit/revert/

khenidak commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -146,14 +146,25 @@ type Store struct {
 	// UpdateStrategy implements resource-specific behavior during updates.
 	UpdateStrategy rest.RESTUpdateStrategy
+	// BeginUpdate is an optional hook that returns a "transaction-like"
+	// commit/cleanup function which will be called at the end of operation,
```
s/commit/revert/

khenidak commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -241,6 +245,63 @@ func (r *GenericREST) defaultOnReadService(service *api.Service) error {
+func (r *GenericREST) beginCreate(ctx context.Context, obj runtime.Object, options *metav1.CreateOptions) (func(context.Context, bool), error) {
+	svc := obj.(*api.Service)
+
+	klog.Errorf("TIM: beginCreate: svc %v/%v - prepare to allocate", svc.Namespace, svc.Name)
```
I just adding an issue here as a reminder to remove or modify these log lines.

khenidak commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -241,6 +245,63 @@ func (r *GenericREST) defaultOnReadService(service *api.Service) error {
+	// it manually. This has to happen here and not in any earlier hooks (e.g.
+	// defaulting) because it needs to be aware of flags and be able to access
+	// API storage.
+	txn, err := r.alloc.allocateCreate(svc, dryrun.IsDryRun(options.DryRun))
```
this is WAY cleaner. thank you!

khenidak commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -241,6 +245,63 @@ func (r *GenericREST) defaultOnReadService(service *api.Service) error {
+	}
+
+	// Our cleanup callback
+	finish := func(_ context.Context, success bool) {
```
might be better to replace success with error variable. Allowing finish functions to log error and call commit/revert. It will show user `this thing was reverted because the op generated this xyz error`

thockin replied:
It originally did, but that made the generic code more complex wrt variable capture.  If API machinery folks want it back, I am happy to do so.

khenidak commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4421,11 +4422,11 @@ func validateServiceExternalTrafficFieldsValue(service *core.Service) field.Erro
 }
 
-// ValidateServiceExternalTrafficFieldsCombination validates if ExternalTrafficPolicy,
+// validateServiceExternalTrafficFieldsCombination validates if ExternalTrafficPolicy,
```
might worth a todo to separate the tests against Create/Update that focuses on ExternalTrafficPolicy into a separate function. The test Create/Update funcs are just too big to comprehend.

thockin replied:
Agree - lots of room for cleanup

thockin committed:
Svc REST: Rename a long, hard function name

thockin committed:
Svc REST: Allow multi-IP-family in tests

thockin committed:
Svc REST: Overhaul Create test wrt dual-stack

This includes a few cases.

1) TestCreateIgnoresIPFamilyForExternalName: Prove that ExternalName is
ignored for dual-stack.  A small set of test cases were chosen to
demonstrate.

2) TestCreateIgnoresIPFamilyWithoutDualStack: Prove that when the
dual-stack gate is off, all services are ignored for dual-stack.  A
small set of test cases were chosen to demonstrate

3) TestCreateInitIPFields: Run over a huge array of test cases for
dual-stack.  This was generated by this program:
https://gist.github.com/thockin/cccc9c9a580b4830ee0946ddd43eeafe and
then updated by hand.

thockin committed:
Svc REST: Dedup tests for defaulting

thockin committed:
Add dry-run support to the IP allocator subsystem

thockin committed:
Svc REST: Set Cluster IPs during dry-run Create

Dry-run should behave like a real API call and return valid results.

thockin committed:
Svc REST: Test that ExternalName doesn't set IPs

thockin committed:
Svc REST: Test that Headless doesn't set IPs

thockin committed:
Svc REST: De-layer Delete

Gut the "outer" Delete() and move it to the inner AfterDelete().

caesarxuchao commented:
> @fedebongio @caesarxuchao no need for anyone to look at this yet except morbid curiosity :)

Feel free to ping me in chat when this is ready for review, though I'm not sure if this is would be more appropriately reviewed by sig-network than sig-apimachinery.

thockin committed:
Svc REST: De-layer Update

This is the last layered method.  All allocator logic is moved to the
beginUpdate() path.  Removing the now-useless layer will happen in a
subsequent commit.

thockin committed:
Svc REST: Move ResourceLocation() to 'inner' layer

Part of the de-layering effort.

Also move the test.

thockin committed:
Svc REST: Remove obviously unused args

This is part of the de-layering conclusion and cleanup.

Bridge the old tests into the new REST.  This will all be removed soon.

thockin committed:
Svc REST: De-layering done! Convert to 1 layer

This is the culmination of all the previous commits which made this last
move less dramatic.  More tests and cleanup commits will follow.

Background, for future archaeologists:

Service has (had) an "outer" and "inner" REST handler.  This is because of how we do IP and port allocations synchronously, but since we don't have API transactions, we need to roll those back in case of a failure.  Both layers use the same `Strategy`, but the outer calls into the inner, which causes a lot of complexity in the code (including an open-coded partial reimplementation of a date-unknown snapshot of the generic REST code) and results in `Prepare` and `Validate` hooks being called twice.

The "normal" REST flow seems to be:

```
mutating webhooks
generic REST store Create {
    cleanup = BeginCreate
    BeforeCreate {
        strategy.PrepareForCreate {
            dropDisabledFields
        }
        strategy.Validate
        strategy.Canonicalize
    }
    createValidation (validating webhooks)
    storage Create
    cleanup
    AfterCreate
    Decorator
}
```

Service (before this series of commits) does:

```
mutating webhooks
svc custom Create {
    BeforeCreate {
        strategy.PrepareForCreate {
            dropDisabledFields
        }
        strategy.Validate
        strategy.Canonicalize
    }
    Allocations
    inner (generic) Create {
        cleanup = BeginCreate
        BeforeCreate {
            strategy.PrepareForCreate {
                dropDisabledFields
            }
            strategy.Validate
            strategy.Canonicalize
        }
        createValidation (validating webhooks)
        storage Create
        cleanup
        AfterCreate
        Decorator
    }
}
```

After this:

```
mutating webhooks
generic REST store Create {
    cleanup = BeginCreate
        Allocations
    BeforeCreate {
        strategy.PrepareForCreate {
            dropDisabledFields
        }
        strategy.Validate
        strategy.Canonicalize
    }
    createValidation (validating webhooks)
    storage Create
    cleanup
    AfterCreate
        Rollback allocations on error
    Decorator
}
```

thockin committed:
Svc REST: Remove old, now unused stubs

aojea commented:
/cc

thockin committed:
Svc REST: Rename service NewGenericREST to NewREST

Just like all the other registries.

aojea commented:
@thockin is anything we can help?

aojea commented:
@thockin are you going to work on this?
I'd like if we can prioritise it  and also solve https://github.com/kubernetes/kubernetes/pull/102955

aojea commented:
legit failure
```
        s: "failed to update Service \"nodeport-service\": Service \"nodeport-service\" is invalid: spec.externalTrafficPolicy: Invalid value: \"Cluster\": ExternalTrafficPolicy can only be set on NodePort and LoadBalancer service",
```

thockin committed:
Svc REST: Add a transaction API

This will be used in upcoming commits, but for easier history and review
it is pretty stand-alone.

thockin committed:
Svc REST: Move allocations in Create into funcs

All the logic remains unchanged, just reorganized.  The functions are
imperfect but emphasize the change being made and can be cleaned up
subsequently.

This makes the following steps easier to comprehend.

aojea commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -95,54 +101,37 @@ func NewREST(
-	serviceIPs ipallocator.Interface,
-	secondaryServiceIPs ipallocator.Interface,
-	serviceNodePorts portallocator.Interface,
+	serviceIPFamily api.IPFamily,
```
`defaultServiceIPFamily`??

thockin committed:
Svc REST: better test checks in new tests

"Has()" was strengthened in the older rest_test, now in the newer.

thockin committed:
Svc REST: Remove redundant Create tests

These cases are all covered in storage_test.

thockin committed:
Svc REST: Better NodePort tests

This commit ports the NodePort test from rest_test to storage_test.
It's not a direct port, though.  I have added many more cases (much more
exhaustive) and more assertions.

This includes cases for gate MixedProtocolLBService.

thockin committed:
Svc REST: Rename some tests for clarity

thockin committed:
Svc REST: Move test to reduce diff in next commits

No changes - just move.

thockin committed:
Svc REST: HealthCheckNodePort tests

This commit ports the ExternalTrafficPolicy and HealthCheckNodePort
tests from rest_test to storage_test.  It's not a direct port, though.
I have added more cases (much more exhaustive) and more assertions.

thockin committed:
Svc REST: IP and port reallocation

Make sure the logic that was covered in rest_test is covered in
storage_test.

thockin committed:
Svc REST: Remove redundant Get test

thockin committed:
Svc REST: Clean up redundant delete tests

thockin committed:
Svc REST: Add new model of feature tests

This scaffolding allows us to assert more on each test case, and more
consistently.

Set input fields from output fields IFF they are expected AND not set on
input.  This allows us to verify the "after" state (expected) whether
the test case specified the value or not, and still pass the generic
cmp.Equal.

Use this in a few tests to prove its worth, more to do.

Some of the existing tests that are focused on create and delete can
probably be replaced by these.

This could be used in other test cases that are open-coding a lot of the
same stuff.  Later commits.

aojea commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -360,6 +360,9 @@ func finishNothing(context.Context, bool) {}
 // Create inserts a new item according to the unique key from the object.
 func (e *Store) Create(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, options *metav1.CreateOptions) (runtime.Object, error) {
+	// Make a copy so we never mutate the passed-in object (e.g. in tests)
+	obj = obj.DeepCopyObject()
```
xref https://groups.google.com/g/kubernetes-sig-api-machinery/c/DZKBRdKIhmk/m/djddw5DgCAAJ

aojea commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -489,9 +485,6 @@ func (rs *REST) Update(ctx context.Context, name string, objInfo rest.UpdatedObj
 		return nil, false, err
 	}
 	externalTrafficPolicyUpdate(oldService, service)
```
This is mutating the service to set `service.Spec.ExternalTrafficPolicy`, 
I think that is the reason why there is a validation here,  should `externalTrafficPolicyUpdate()` be moved to `strategy.PrepareForUpdate()`?

thockin replied:
good catch!

thockin replied:
made a new commit just for this

thockin replied:
Put new commit at the front - it can be merged independent of the rest.

aojea commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -182,71 +182,68 @@ func (rs *REST) Create(ctx context.Context, obj runtime.Object, createValidation
-			// these should be caught by an eventual reconciliation / restart
-			utilruntime.HandleError(fmt.Errorf("error(s) committing service node-ports changes: %v", el))
+	// Allocate ClusterIPs
+	//FIXME: we need to put values in, even if dry run - else validation should
```
xref: https://groups.google.com/g/kubernetes-sig-api-machinery/c/_-5TKHXHcXE/m/RfKj7CtzAQAJ

aojea commented on pkg/registry/core/service/storage/transaction.go:
```diff
@@ -0,0 +1,62 @@
+limitations under the License.
+*/
+
+package storage
```
Reminder, this package needs unit tests and it can be easy to work on them in parallel

aojea replied:
the promised test, these closures things, I didn't expect this test to work this way :smile: 
```go
/*
Copyright 2020 The Kubernetes Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package storage

import (
	"testing"
)

func Test_metaTransaction(t *testing.T) {
	const initial = 10
	var temp int

	tests := []struct {
		name  string
		mt    metaTransaction
		start int
		want  int
	}{
		{
			name: "commit and revert match",
			mt: metaTransaction{
				callbackTransaction{
					commit: func() {
						temp = temp + 1
					},
					revert: func() {
						temp = temp - 1
					},
				},
			},
			want: 10,
		},
		{
			name: "commit and revert match multiple times",
			mt: metaTransaction{
				callbackTransaction{
					commit: func() {
						temp = temp + 1
					},
					revert: func() {
						temp = temp - 1
					},
				},
				callbackTransaction{
					commit: func() {
						temp = temp + 2
					},
					revert: func() {
						temp = temp - 2
					},
				},
				callbackTransaction{
					commit: func() {
						temp = temp + 3
					},
					revert: func() {
						temp = temp - 3
					},
				},
			},
			want: 10,
		},
		{
			name: "missing revert",
			mt: metaTransaction{
				callbackTransaction{
					commit: func() {
						temp = temp + 1
					},
					revert: func() {
						temp = temp - 1
					},
				},
				callbackTransaction{
					commit: func() {
						temp = temp + 2
					},
				},
				callbackTransaction{
					commit: func() {
						temp = temp + 3
					},
					revert: func() {
						temp = temp - 3
					},
				},
			},
			want: 12,
		},
		{
			name: "missing commit",
			mt: metaTransaction{
				callbackTransaction{
					commit: func() {
						temp = temp + 1
					},
					revert: func() {
						temp = temp - 1
					},
				},
				callbackTransaction{
					revert: func() {
						temp = temp - 2
					},
				},
				callbackTransaction{
					commit: func() {
						temp = temp + 3
					},
					revert: func() {
						temp = temp - 3
					},
				},
			},
			want: 8,
		},
		{
			name: "commit and revert match multiple but different order",
			mt: metaTransaction{
				callbackTransaction{
					commit: func() {
						temp = temp + 1
					},
					revert: func() {
						temp = temp - 2
					},
				},
				callbackTransaction{
					commit: func() {
						temp = temp + 2
					},
					revert: func() {
						temp = temp - 1
					},
				},
				callbackTransaction{
					commit: func() {
						temp = temp + 3
					},
					revert: func() {
						temp = temp - 3
					},
				},
			},
			want: 10,
		},
	}
	t.Parallel()
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			temp = initial
			tt.mt.Commit()
			tt.mt.Revert()
			if temp != tt.want {
				t.Fatalf("expected %d got %d", tt.want, temp)
			}
		})
	}
}
```

thockin replied:
Thanks!  I did forget.  Added in a FIXUP commit

aojea commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -1158,6 +1176,44 @@ func allocateHealthCheckNodePort(service *api.Service, nodePortOp *portallocator
 
+//FIXME: rename and merge with initNodePorts
+func (al *RESTAllocStuff) allocServiceNodePortsNew(service *api.Service, dryRun bool) (transaction, error) {
+	// The allocator tracks dry-run-ness internally.
```
I think we can exit fast and avoid the nested if
```
if service.Spec.Type != api.ServiceTypeNodePort && service.Spec.Type != api.ServiceTypeLoadBalancer {
 return nil, nil
}
}

thockin replied:
If we allow nil `transaction` then we have to check for nil everywhere.  We could add a `nullTransaction` but I don't think it will be used more than this spot...

aojea commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -1158,6 +1176,44 @@ func allocateHealthCheckNodePort(service *api.Service, nodePortOp *portallocator
+
+	// TODO: This creates nodePorts if needed. In the future nodePorts may be cleared if *never* used.
+	// But for now we stick to the KEP "don't allocate new node ports but do not deallocate existing node ports if set"
+	if service.Spec.Type == api.ServiceTypeNodePort || service.Spec.Type == api.ServiceTypeLoadBalancer {
```
if we exit fast we don't need this https://github.com/kubernetes/kubernetes/pull/96684/commits/75c3780f840b0fcfa8875a71a12628c3bfbd2bec#r668076112

thockin replied:
I don't even know what this comment means, frankly.  I'm going to remove it.   It's not an actionable TODO.

thockin committed:
Svc: Move ETP clearing to dropTypeDependentFields

I  am not sure why ExternalTrafficPolicy was different, but this is more
consistent with other field clearing logic.

thockin committed:
Svc REST: Add a delete-with-finalizer test

This is a long-standing bug that gets fixed "for free" in the
de-layering.

thockin committed:
Svc REST: Use "prove" helpers in other tests

aojea commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -390,13 +371,12 @@ func (rs *REST) healthCheckNodePortUpdate(oldService, service *api.Service, node
 		klog.V(4).Infof("Releasing healthCheckNodePort: %d", oldHealthCheckNodePort)
 		nodePortOp.ReleaseDeferred(int(oldHealthCheckNodePort))
-		// Clear the HealthCheckNodePort field.
-		service.Spec.HealthCheckNodePort = 0
```
is ok to remove it completely or should we have something on strategy clearing the healthchecknodeport?

aojea replied:
it is already cleared in `dropTypeDependentFields` , we can resolve this comment

thockin committed:
Fix validation on ETP: "" is not valid

This was causing tests to pass which ought not be passing.  This is not
an API change because we default the value of it when needed.  So we
would never see this in the wild, but it makes the tests sloppy.

thockin committed:
Svc REST: Change ETP create test to a feature test

All the same test cases and more.

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
+	svc         *api.Service
+	expectError bool
+
+	// We could divine these by looking at the Service, but that's a
```
divine? as in identify?

thockin replied:
meaning "figure out" or "discover".  Will reword.

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
+	}
+	for _, clip := range clips {
+		if !ipIsAllocated(t, storage.alloc.serviceIPAllocatorsByFamily[familyOf(clip)], clip) {
+			t.Errorf("%s: expected clusterIP to be allocated: %q", callName(before, after), clip)
```
We can save alot of code by just having `t.Fatalf()/t.Errorf()` called inside helpers such as `ipIsAllocated` -- should remove a lot of `if statement cruft in testing`- I am not against how it is done now. Just an improvement.

thockin replied:
Almost all calls to `ipIsAllocated` will get wrapped in `prove...` calls in later commits.  Changing this would cause a lot of rebase work for no net gain once all commits are applied.

OK to decline?

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
+		}
+	}
+
+	for i, fam := range after.Spec.IPFamilies {
```
This is assuming that `IPFamilies` is of correct length. Should we have this assumption in test?

thockin replied:
good point.  Will assert same len

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
+func proveClusterIPsDeallocated(t *testing.T, storage *GenericREST, before, after *api.Service) {
+	t.Helper()
+
+	if after != nil && after.Spec.ClusterIP != api.ClusterIPNone {
```
bailout early. by having
```
if after != nil && after.Spec.ClusterIP == api.ClusterIPNone {
return
}
```
instead of the double ifs guarding both assertion.

thockin replied:
but I don't want to return - I want to check before, too.  Maybe I miss your point?

khenidak replied:
We can ignore this for now.

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
+	if before != nil && before.Spec.ClusterIP != api.ClusterIPNone {
+		clips := []string{}
+		if utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
+			clips = before.Spec.ClusterIPs
```
clipping IPs by gate is a repeated call. maybe a helper func?

thockin replied:
We only do it in 2 places I think, and once the gate goes away that whole block goes away (no need for a helper), so I'd rather ride it out, ok?

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
+		}
+	}
+
+	if before != nil {
```
Commit `Service REST: feature tests`: 
Correctness of struct before correctness of allocators. e.g. 

```
if before != nil ... go through wants/got check
then
for _,  p:= range after.Spec.Ports {
   is port allocated check
}
```
The pattern is also repeated else where.

thockin replied:
Yeah, largely because it used to only have "after" then I added "before" :)  Note that the blocks are looking at different objects, so it's not a nil deref.  I kind of prefer it this way because it seems more obvious to me that we check the current state first and foremost.  convinced?

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
+			if !verifyEquiv(t, "update", &tc.update, updatedSvc) {
+				return
+			}
+			if tc.update.expectClusterIPs {
```
Commit `Service REST: feature tests`: 
It is basically the same call as L6051 define an inlined func and let us call it twice.

thockin replied:
ACK.

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
+			}
+
+			// Delete the object and check the results.
+			_, _, err = storage.Delete(ctx, tc.update.svc.Name, rest.ValidateAllObjectFunc, &metav1.DeleteOptions{})
```
Commit `Service REST: feature tests`: 
We are deferring a Delete() call L6043. And we are deferring ...Store.Destroy() -- Let us keep the destroy, remove the deferred delete. It just lengthen the tests duration.

thockin replied:
The defer'ed delete is to cover what happens if we abort the test early (`Fatalf()`).  The only way to avoid that need is to move the storage into the Run() func, and that makes the test REALLY slow when there are lots of cases

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
+			expectClusterIPs: true,
+			expectNodePorts:  true,
+		},
+		update: svcTestCase{
```
Commit `Service REST: feature tests`: 
Question: Do we expect to reset LB ingress address slice when the user does that conversion? (This is an out of band question)

thockin replied:
yes, but I don't think we set it at all until a controller allocates an IP, so it will always be nil here.

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
+	helpTestCreateUpdateDelete(t, testCases)
+}
+
+func TestFeatureType(t *testing.T) {
```
Commit `Service REST: feature tests`: 
Missing `headless` to and from tests.

khenidak replied:
I just realized that TestFeatureClusterIPs() does headless->clusterIPs but not the other way around.

thockin replied:
I am confused - it does test headless here...

Will add case in ClusterIPs

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
+					s.Spec.SessionAffinityConfig = nil
+				}),
+			expectClusterIPs: true,
+		},
```
Commit `Service REST: feature tests`: 
SessionAffinity is immutable right? Shouldn't we add a test With-SessionAffinity=>Without-SessionAffinity and test clusterIPs have not been released?

thockin replied:
It's not immutable.  These tests flip it from None->ClietIP and back.

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
+		update: svcTestCase{
+			svc: svctest.MakeService("foo", svctest.SetTypeClusterIP,
+				svctest.SetHeadless),
+			expectError: true,
```
Commit `Service REST: feature tests`: 
this is a valid use case. e.g. converting headful to headless. no?

thockin replied:
nope.  Can't modify ClusterIP.

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
+			expectClusterIPs: true,
+		},
+	}, {
+		name: "ClusterIP_selector:valid:empty",
```
Commit `Service REST: feature tests`: 
test case `ClusterIP_selector:valid:nil` and `ClusterIP_selector:valid:empty` are the same.

thockin replied:
good catch

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
+			expectClusterIPs: true,
+		},
+	}, {
+		name: "ClusterIP_selector:empty:valid",
```
Commit `Service REST: feature tests`: 
is this really needed? We are not testing anything here.

thockin replied:
just proving it can be changed.

khenidak commented:
Added comments for `Service REST: feature tests` commit. 


All commits are titled with
Commit `<commit title>`:

e.g.
Commit `Service REST: feature tests`: for all my comments on that commit.

khenidak commented on pkg/registry/core/service/strategy.go:
```diff
@@ -305,11 +304,12 @@ func (serviceStatusStrategy) WarningsOnUpdate(ctx context.Context, obj, old runt
 // allocate one.  If they resubmit that same YAML, we want it to succeed.
-func patchAllocatedValues(newSvc, oldSvc *api.Service) {
+//FIXME: move this to pkg/registry/core/service/storage
+func PatchAllocatedValues(newSvc, oldSvc *api.Service) {
```
Commit `De-layer Service REST: Update`:
No test covers this func.

thockin replied:
huh.  New commit added.

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5651,6 +5651,109 @@ func TestDeleteTypes(t *testing.T) {
+	createdSvc := obj.(*api.Service)
+
+	// Prove everything was allocated.
+	_, err = storage.Get(ctx, svcName, &metav1.GetOptions{})
```
Commit `Service REST: add a delete-with-finalizer test`:
`createdSvc, err :=  storage.Get(..)`

thockin replied:
ACK

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5651,6 +5651,109 @@ func TestDeleteTypes(t *testing.T) {
+	if err != nil {
+		t.Fatalf("unexpected error getting service: %v", err)
+	}
+	for i, fam := range createdSvc.Spec.IPFamilies {
```
commit `Service REST: add a delete-with-finalizer test`:
this is the same code as L8685-L5609 wrap in a inlined func and call it twice?

thockin replied:
ACK

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5651,6 +5651,109 @@ func TestDeleteTypes(t *testing.T) {
+	if err == nil {
+		t.Fatalf("unexpected success getting service")
+	}
+	for i, fam := range createdSvc.Spec.IPFamilies {
```
commit `Service REST: add a delete-with-finalizer test`:
we really need to wrap these things for alloc/dealloc.

thockin replied:
That happens in a subsequent commit :)

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5593,6 +5593,64 @@ func TestCreateDryRun(t *testing.T) {
+		svc: svctest.MakeService("foo",
+			svctest.SetTypeExternalName),
+	}, {
+		name: "type:ClusterIP",
```
Commit `Service REST test: clean up redundant delete tests`:
Missing duals-tack with two IPs to test allocator didn't not mess up and just freed on ClusterIP

thockin replied:
This whole test should be removed after the feature tests prove create-update-delete more completely.  I'll leave this as-is here and remove it in the later commit

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5593,6 +5593,64 @@ func TestCreateDryRun(t *testing.T) {
+				t.Fatalf("unexpected error deleting service: %v", err)
+			}
+			if !deleted {
+				t.Fatalf("expected service to be deleted")
```
Commit `Service REST test: clean up redundant delete tests`:
are we not testing deallocation here?

thockin replied:
removed in a later commit in favor of the common CUD tests.

khenidak commented:
General Note:
commits
1. De-layer Service REST: Delete
2. Service REST: add a delete-with-finalizer test
3. Service REST test: clean up redundant delete tests

**can be squashed together.**

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -5580,3 +5592,81 @@ func TestCreateDryRun(t *testing.T) {
+			}
+			for _, p := range createdSvc.Spec.Ports {
+				if !portIsAllocated(t, storage.alloc.serviceNodePorts, p.NodePort) {
+					t.Errorf("expected port to still be allocated: %v", p.NodePort)
```
Commit `De-layer Service REST: Delete`:
`%+v` ? (similarly other struct types)

thockin replied:
ACK, will try to use %q for IPs and %d for ports.

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -4787,10 +4797,24 @@ func TestCreateReallocation(t *testing.T) {
 			svc2.Spec.ClusterIPs = createdSvc.Spec.ClusterIPs
 			svc2.Spec.Ports = createdSvc.Spec.Ports
 
+			// Create again
```
Commit `De-layer Service REST: Delete`
I know it might be an overkill. But i would like to add change service name in the second create. It is more likely the scenario that people will use and most importantly i would like to confuse the allocator/repair loop.

thockin replied:
ACK

khenidak commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -250,6 +266,33 @@ func (r *GenericREST) defaultOnReadService(service *api.Service) {
+func (r *GenericREST) afterDelete(obj runtime.Object, options *metav1.DeleteOptions) {
+	svc := obj.(*api.Service)
+
+	klog.Errorf("TIM: afterDelete: svc %v/%v - prepare to deallocate", svc.Namespace, svc.Name)
```
Commit `De-layer Service REST: Delete`:
it does not have a FIXME. Reminder to remove.

khenidak commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -250,6 +266,33 @@ func (r *GenericREST) defaultOnReadService(service *api.Service) {
+			klog.Errorf("delete service endpoints %s/%s failed: %v", svc.Name, svc.Namespace, err)
+		}
+
+		klog.Errorf("TIM: afterDelete: svc %v/%v - deallocating", svc.Namespace, svc.Name)
```
Commit `De-layer Service REST: Delete`:
it does not have a FIXME. Reminder to remove.

khenidak commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -250,6 +266,33 @@ func (r *GenericREST) defaultOnReadService(service *api.Service) {
+		// endpoints if a new service is created with the same name
+		_, _, err := r.endpoints.Delete(ctx, svc.Name, rest.ValidateAllObjectFunc, &metav1.DeleteOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			klog.Errorf("delete service endpoints %s/%s failed: %v", svc.Name, svc.Namespace, err)
```
Commit `De-layer Service REST: Delete`:
Comment is un-actionable. We should append `endpoint controller` should clean this up eventually.  

Also: Why are we deleting endpoints and not endpointSlice?

thockin replied:
This is a vestigial thing.  We PROBABLY can get rid of this and let the controller do it, but there was some fear that one might reuse a service name and get the old endpoints (race).  I agree we should either do both or do neither.  Will add a TODO

khenidak commented:
Commits:
1. Move test to reduce diffs in next commits
2. Service REST test: Remove redundant Get test
3. WIP: Service: Use DeepCopy() and fix tests 

have been reviewed with no comments

thockin commented:
checkpoint: All Kal's feedback in commit `Service REST: feature tests` is amended to that commit.  tests pass.  Pushed

thockin committed:
Svc REST: Add a test for PatchAllocatedValues

thockin commented:
To the best of my knowledge all Kal's comments are resolved.  Force-pushing now.  Will run a sweep to see what tests fail on intermediate commits.  Still more tests to write and port, and plenty of FIXME.

thockin committed:
Svc REST: Set Cluster IPs during dry-run Update()

Dry-run should return valid results.

Also add a test.

thockin commented:
implemented and added a test for update dry run

aojea commented:
/retest

aojea commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -240,3 +291,163 @@ func (r *GenericREST) defaultOnReadService(service *api.Service) {
+		// TODO: This is clumsy.  It was added for fear that the endpoints
+		// controller might lag, and we could end up rusing the service name
+		// with old endpoints.  We should solve that better and remove this, or
+		// else we should do this for EndpointSlice, too.
```
I don't think slices has this problem, since it reconciles on slices events, when the controller restarts it will requeue the orphan endpoint and it should be deleted ... I didn't verify it though

aojea commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -240,3 +291,163 @@ func (r *GenericREST) defaultOnReadService(service *api.Service) {
+	}
+
+	// If a port *number* was specified, find the corresponding service port name
+	if portNum, err := strconv.ParseInt(portStr, 10, 64); err == nil {
```
this was this way before, but it seems that if `err != nil` ... it swallows it

thockin replied:
If it doesn't parse as an int, we treat it as a name.  I think that's correct, no?

aojea replied:
ah right

thockin committed:
Svc REST: Beef up NodePort tests

Remove old test from rest_test.go.

aojea commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -361,6 +361,9 @@ func finishNothing(context.Context, bool) {}
 // Create inserts a new item according to the unique key from the object.
 func (e *Store) Create(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, options *metav1.CreateOptions) (runtime.Object, error) {
+	// Make a copy so we never mutate the passed-in object (e.g. in tests)
+	obj = obj.DeepCopyObject()
```
this commit is ... controversial https://github.com/kubernetes/kubernetes/pull/103545#issuecomment-895457403

thockin replied:
yes, I will back-populate whatever we decide there :)

thockin committed:
Svc REST: Fix single<->dual-stack updates

This removes the old rest_tests and adds significantly more coverage.
Maybe too much.  The v4,v6 and v6,v4 tables are identical but for the
order of families.

This exposed that `trimFieldsForDualStackDowngrade` is called too late
to do anything (since we don't run strategy twice any more).  I moved
similar logic to `PatchAllocatedValues` but I hit on some unclarity.

Specifically, consider a PATCH operation.

Assume I have a valid dual-stack service (with 2 IPs, 2 families, and
policy either require or prefer). What fields can I patch, on their own,
to trigger a downgrade to single-stack?

I think patching policy to "single" is pretty clear intent.

But what if I leave policy and only patch `ipFamilies` down to a single
value (without violating the "can't change first family" rule)?

Or what if I patch `clusterIPs` down to a single IP value?

After much discussion, we decided to make a small API change (OK since
we are beta).  When you want a dual-stack Service you MUST specify the
`ipFamilyPolicy`.  Now we can infer less and avoid ambiguity.

thockin committed:
Svc REST: Add proof funcs in feature test logic

Allows for more control of tests to assert specific things.

thockin committed:
Svc REST: add a beforeUpdate hook in feature tests

thockin committed:
Svc REST: allow tests to set cluster IP families

thockin committed:
Svc REST: Better errors on stack-downgrades

Converting dual-stack to single-stack needs good errors.

thockin committed:
Svc REST: Validate input before IP allocation

This commit started as removing FIXME comments, but in doing so I
realized that the IP allocation process was using unvalidated user
input.  Before de-layering, validation was called twice - once before
init and once after, which the init code depended on.

Fortunately (or not?) we had duplicative checks that caught errors but
with less friendly messages.

This commit calls validation before initializing the rest of the
IP-related fields.

This also re-organizes that code a bit, cleans up error messages and
comments, and adds a test SPECIFICALLY for the errors in those cases.

thockin committed:
Svc REST: Make ipFamilyPolicy authoritative

Previously we would try to infer the `ipFamilyPolicy` from `clusterIPs`
and/or `ipFamilies`.  That is too tricky.  Now you MUST specify
`ipFamilyPolicy` as one of the dual-stack options in order to get a
dual-stack service.

thockin committed:
Svc REST: Beef up ports test, remove old form

thockin committed:
Svc REST: Remove overlapping rest_tests

Most are moved to storage_test

thockin committed:
Svc REST: Add InternalTrafficPolicy tests

Remove older form.

thockin committed:
Svc REST: Remove old rest_test

All the tests have been ported to storage_test.go

khenidak commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -845,8 +872,46 @@ func (al *RESTAllocStuff) initIPFamilyFields(oldService, service *api.Service) e
+		// If they didn't specify policy, or specified anything but
+		// single-stack AND they reduced these fields, it's an error.  They
+		// need to specify policy.
+		if np == nil || *np != api.IPFamilyPolicySingleStack {
```
Ok. `on update policy must be set` That means on Post/Patch/etc the user must either `NOT touch it` or `*DELIBERATELY change it*`. that is the agreement right? so:
```go
if !isUpdated(old, new) /* checks if old != nil *{
 // split the func in two: one for create and one for update
 // call create logic
}
// ASSUMING THAT HEADLESS AND EXTERNAL NAME HAS BEEN HANDLED
// the rest of function is for update case only
 np := service.Spec.IPFamilyPolicy
 if np == nil{ 
   // error immutable once set.
 }
 // single stack case
 if np == api.IPFamilyPolicySingleStack /*we don't care what old was here*/{
 // the below should never fail since it is 
 //1) it is on update path  [i.e.,must have something set prior]
 //2) !ExternalName
 //3) !Headless
 service.Spec.ClusterIPs = service.Spec.ClusterIPs[:1]
 service.Spec.IPFamilies = service.Spec.IPFamilies[:1]
 
 // return here
 }
 // at this stage we know np is not nil - we also know it is not SingleStack
if reducedClusterIPs(oldService, service) {...}
if reducedIPFamilies(oldService, service) {...}
if len(el) > 0 {.. return errors ...}
return here
}
// the only remaining logic is testing immutability of pre allocated ClusterIPs. both are immutable once set.
```

thockin committed:
DeepCopy() input objects in Service REST test

Since the PR to do this deeper in the stack was declined, we'll do it
ourselves.  This ensures that we don't accidentally mutate the input and
then compare that mutated form to the result (which caused previous test
failures).

thockin committed:
Svc REST: Fix comments to make next commits easier

khenidak commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -248,6 +247,9 @@ func (al *RESTAllocStuff) allocClusterIPs(service *api.Service, toAlloc map[api.
 			allocated[family] = allocatedIP.String()
 		} else {
 			parsedIP := netutils.ParseIPSloppy(ip)
+			if parsedIP == nil {
```
We don't need that. It is parsed as part of validation: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L6361

Does this gets called before Validation?

thockin replied:
I think we do a disservice by not preparing for bugs.  The "None" bug slipped through because we don't handle this "can't happen" case which did happen.  I'd like to see us handle more such cases as "WTF?!?" errors rather than crash or silently ignore

khenidak commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -269,11 +271,14 @@ func (al *RESTAllocStuff) releaseClusterIPs(toRelease map[api.IPFamily]string) (
 		}
 
 		parsedIP := netutils.ParseIPSloppy(ip)
+		if parsedIP == nil {
```
same as L250

khenidak commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -157,7 +158,21 @@ func newStorageWithPods(t *testing.T, ipFamilies []api.IPFamily, pods []api.Pod,
 		t.Fatalf("unexpected error from REST storage: %v", err)
 	}
-	return serviceStorage, statusStorage, server
+	return &TestREST{serviceStorage}, statusStorage, server
```
do we need `TestREST` to be exportable?

thockin replied:
Good point.  Will rename and make it more obvious.

khenidak reviewed (commented):
one minor issue.

thockin commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -845,8 +872,46 @@ func (al *RESTAllocStuff) initIPFamilyFields(oldService, service *api.Service) e
 		return nil // nothing more to do.
 	}
 
+	// Update only prep work.
```
I don't know what you mean here - This correctly handles headless, I think

thockin commented:
I have tried to address comments, but some are obsolete with the thrashing the last few days.

I have edited the TODO list in the first comment on this PR to defer a bunch of things, but you can see my punch-list there.

thockin committed:
Svc REST: Move normalizeClusterIPs to storage pkg

All the meaningful callers of it are in that pkg anyway.  Removes some
FIXMEs.

thockin committed:
Svc REST: Move patchAllocatedValues to storage pkg

All the meaningful callers of it are in that pkg anyway.  Removes 1
FIXME.

thockin committed:
Svc REST: Move tests and scaffolding around

No code edits.  Just a little whitespace, adding comments, and
re-ordering functions.

thockin committed:
Svc REST: rename allocServiceClusterIPsNew

thockin committed:
Svc REST: rename allocClusterIPs -> allocIPs

thockin committed:
Svc REST: rename allocServiceNodePortsNew

thockin committed:
Svc REST: rename releaseClusterIPs -> releaseIPs

thockin committed:
Svc REST: rename allocUpdateServiceClusterIPsNew

thockin committed:
Svc REST: rename allocUpdateServiceNodePortsNew

thockin committed:
Svc REST: rename handleClusterIPsForUpdatedService

thockin committed:
Svc REST: rename healthCheckNodePortUpdate

thockin committed:
Svc REST: rename allocateHealthCheckNodePort

thockin committed:
Svc REST: rename allocServiceClusterIP

thockin committed:
Svc REST: rename allocServiceClusterIPs

thockin committed:
Svc REST: rename releaseServiceClusterIP

thockin committed:
Svc REST: rename releaseServiceClusterIPs

thockin committed:
Svc REST: Make allocHCNP a method: more consistent

thockin committed:
Svc REST: rename rest.go -> alloc.go

thockin committed:
Svc REST: Move alloc code around

This is detrimental to future `blame` but makes it so much morereadable
I convinced myself it was worthwhile.

thockin committed:
Svc REST: Make update/releaseNodePorts methods

More consistent overall.

thockin committed:
Svc REST: Move isValidAddress to storage.go

thockin committed:
Svc REST: Rename GenericREST -> REST

This is consistent with every other registry.  Service is no longer the
oddball.

thockin committed:
Svc REST: Use types for safer arg ordering

In all the places we pass (old, new) or (new, old), use wrapper-types to
make sure that we don't flip the order by accident.

thockin committed:
Svc REST: Rename RESTAllocStuff

This was a dumb placeholder name.

aojea commented:
/retest

thockin committed:
Svc REST: Convert FIXME to TODO

khenidak commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -824,41 +824,35 @@ func (al *RESTAllocStuff) initIPFamilyFields(oldService, service *api.Service) e
+	//     because a PUT that changes clusterIPs from 2 to 1 value but doesn't
+	//     specify ipFamily would work.
+	//   * Update (PATCH): If they didn't specify a policy it will adopt the
+	//     policy from before.
```
nit `adapt` not `adopt`

thockin replied:
I think I meant adopt here - as in "we'll take it as our own"

khenidak replied:
ah. That makes sense. thanks

khenidak commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -824,41 +824,35 @@ func (al *RESTAllocStuff) initIPFamilyFields(oldService, service *api.Service) e
+	//   * Update (PATCH): If they didn't specify a policy it will adopt the
+	//     policy from before.
+	if service.Spec.IPFamilyPolicy == nil {
+		if service.Spec.ClusterIP == api.ClusterIPNone && len(service.Spec.Selector) == 0 {
```
```golang
// if we have old then it prevails against any defaulting (as per comments block earlier)
if oldService != nil && oldService.Spec.IPFamilyPolicy != nil {
// copy from old bail out
}

// there is no oldService || oldService.Spec.IPFamily ==nil
if service.Spec.ClusterIP == api.ClusterIPNone && len(service.Spec.Selector) == 0{
// set as require dual stack and bail out
}
// last case Create (with nil policy) or Update (with no prior policy set because it was externalName) 
// set as single stack and bail out.
```

thockin replied:
ACK

khenidak commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -871,6 +865,31 @@ func (al *RESTAllocStuff) initIPFamilyFields(oldService, service *api.Service) e
+			if sameClusterIPs(oldService, service) && len(service.Spec.ClusterIPs) > 1 {
+				service.Spec.ClusterIPs = service.Spec.ClusterIPs[0:1]
+			}
+			if sameIPFamilies(oldService, service) && len(service.Spec.IPFamilies) > 1 {
```
worth commenting here that ClusterIP[0] is immutable. hence it will always be the correct family.

thockin replied:
ACK

khenidak commented on pkg/registry/core/service/strategy_test.go:
```diff
@@ -1030,30 +1030,6 @@ func TestPatchAllocatedValues(t *testing.T) {
 			svctest.SetTypeClusterIP),
 		expectSameClusterIPs: true,
-	}, {
-		name: "first_IP_patched_because_policy",
```
I couldn't tell. But i assume these tests moved somewhere else?

thockin replied:
These specific cases go away because we later truncate - it made more sense in the same code that does all the other IP logic

khenidak commented:
@thockin we probably want to leave a release note re authoritative nature of `IPFamilyPolicy`.

aojea commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -182,71 +182,68 @@ func (rs *REST) Create(ctx context.Context, obj runtime.Object, createValidation
-		// no clusterips to release
-		toReleaseClusterIPs = nil
+	// Allocate ports
+	if txn, err := al.allocServiceNodePortsNew(service, dryRun); err != nil {
```
`if !dryRun {` ??

aojea replied:
nevermind, is inside

thockin replied:
We still want to assign a value in dryRun, we just don't want to actually allocate it.  That's handled further down-stack.

thockin commented:
Rebased, no net change (the only conflict was ultimately removed)

aojea commented on pkg/api/service/testing/make.go:
```diff
@@ -113,6 +114,26 @@ func MakeServicePort(name string, port int, tgtPort intstr.IntOrString, proto ap
+// SetSelector sets the service selector.
+func SetSelector(sel map[string]string) Tweak {
+	return func(svc *api.Service) {
+		svc.Spec.Selector = sel
```
are we fine having selector as a reference, or do we want to protect from undesired mutations like with the svc.Deepcopy()

aojea replied:
it is fine, all the tests are using it as
```
pkg/registry/core/service/storage/storage_test.go:                              svctest.SetSelector(map[string]string{"k2": "v2"}),
pkg/registry/core/service/storage/storage_test.go:                              svctest.SetSelector(map[string]string{"k2": "v2"}),
```

thockin replied:
huh.  Good question.  I expect it to mostly be used as `SetSelector(map[string]string{"foo": "bar")), so copying is not important but I could also see wasting time on this in the future.  Will fix in a FIXUP commit.

aojea commented on pkg/registry/core/service/storage/storage_test.go:
```diff
@@ -614,3 +616,4562 @@ func TestServiceDefaulting(t *testing.T) {
+}
+
+// Prove that create ignores IPFamily stuff when type is ExternalName.
+func TestCreateIgnoresIPFamilyForExternalName(t *testing.T) {
```
note for others, this disappears later or at least I can't find it

thockin replied:
renamed to `TestCreateIgnoresIPsForExternalName`

aojea commented:
/test

aojea commented:
/test pull-kubernetes-e2e-ipvs-azure-dualstack
/test pull-kubernetes-e2e-iptables-azure-dualstack
/test pull-kubernetes-e2e-kind-dual-canary
/test pull-kubernetes-e2e-kind-ipvs-dual-canary

aojea commented:
@uablrek do you mind trying this PR, it should be ready to merge soon, but is an important change and you use to find corner cases that are not covered by tests :1st_place_medal:

aojea commented:
we 2 e2e tests failing
```
Kubernetes e2e suite: [sig-network] [Feature:IPv6DualStack] should create service with ipv6,v4 cluster ip 

  Service "ipv6ipv4clusterip" is invalid: spec.ipFamilyPolicy: Invalid value: "SingleStack": must be 'RequireDualStack' or 'PreferDualStack' when multiple IP families are specified
```
and
```
Kubernetes e2e suite: [sig-network] [Feature:IPv6DualStack] should create service with ipv4,v6 cluster ip 
 Service "ipv4ipv6clusterip" is invalid: spec.ipFamilyPolicy: Invalid value: "SingleStack": must be 'RequireDualStack' or 'PreferDualStack' when multiple IP families are specified
```

thockin commented:
Are those new e2e?  It was passing.

thockin commented:
All comments (so far) addressed in FIXUP commits (those can squash when it's all done)

aojea commented on test/e2e/network/dual_stack.go:
```diff
@@ -367,7 +367,7 @@ var _ = common.SIGDescribe("[Feature:IPv6DualStack]", func() {
 		expectedPolicy := v1.IPFamilyPolicyRequireDualStack
 		expectedFamilies := []v1.IPFamily{v1.IPv4Protocol, v1.IPv6Protocol}
 
-		service := createService(t.ServiceName, t.Namespace, t.Labels, nil, expectedFamilies)
```
these tests only run on dual-stack clusters, they are not running automatically in presubmits, that is why we didn't hit it.
They run periodically though, we should be able to catch it later

thockin replied:
Aha!

aojea commented:
/lgtm
/hold

holding so Kal can sign off

My compliments and my admiration for this PR, it was really needed :clap: :clap: :clap:

khenidak commented on pkg/registry/core/service/storage/alloc.go:
```diff
@@ -0,0 +1,1101 @@
+			toRelease[al.defaultServiceIPFamily] = oldService.Spec.ClusterIP
+		} else {
+			// dual stack is enabled, collect ClusterIPs by families
+			for i, family := range oldService.Spec.IPFamilies {
```
This is the bug that https://github.com/kubernetes/kubernetes/pull/104742 solves. The old service might be headless and this will yield into problems. Since the code is so different maybe we should just fix here as well (the fix PR has an integration test that will guard against this, but it is not merged yet so there is that)

khenidak replied:
Please ignore this. it is fixed here: https://github.com/kubernetes/kubernetes/pull/96684/files/0dcc4e7193f744124c6afa650df0f5fb6921dfc8#diff-989117d719398a16288c815c08122ba373274bbfebc880bb78b67e405db21fd9R670

aojea commented:
/label tide/merge-method-squash

khenidak commented on pkg/registry/core/service/strategy_test.go:
```diff
@@ -115,15 +114,6 @@ func makeValidServiceCustom(tweaks ...func(svc *api.Service)) *api.Service {
-}
-
 // TODO: This should be done on types that are not part of our API
 func TestBeforeUpdate(t *testing.T) {
```
we no longer need this right?

thockin replied:
Removed in a new FIXUP commit.

khenidak reviewed (commented):
fewer nits... so close!

khenidak commented:
/retest

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4506,22 +4507,58 @@ func validateServicePort(sp *core.ServicePort, requireName, isHeadlessService bo
+				"may only be set when `type` is 'LoadBalancer' and `externalTrafficPolicy` is 'Local'"))
+		}
+	} else {
+		if service.Spec.HealthCheckNodePort == 0 {
```
`IsValidPortNum()` a few lines later checks for negative

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4545,29 +4582,6 @@ func validateServiceInternalTrafficFieldsValue(service *core.Service) field.Erro
 	return allErrs
 }
 
-// ValidateServiceExternalTrafficFieldsCombination validates if ExternalTrafficPolicy,
```
I have more where that came from - just wait for the next one!

thockin commented on pkg/registry/core/service/ipallocator/allocator_test.go:
```diff
@@ -514,3 +514,69 @@ func expectMetrics(t *testing.T, label string, em testMetrics) {
+			if err := r.DryRun().Release(cidr.IP); err != nil {
+				t.Fatalf("unexpected failure: %v", err)
+			}
+			if u := r.Used(); u != baseUsed {
```
ACK - in next push

thockin commented on pkg/registry/core/service/storage/alloc.go:
```diff
@@ -0,0 +1,1101 @@
+		return nil, err
+	}
+
+	// Allocate ClusterIPs
```
Yes.  I have half a patch to fix it already :)

thockin commented on pkg/registry/core/service/storage/alloc.go:
```diff
@@ -0,0 +1,1101 @@
+	//does (and is tested!).  Fixing that all is a big PR and will have to
+	//happen later.
+	if txn, err := al.txnAllocClusterIPs(service, dryRun); err != nil {
+		result.Revert()
```
I put it there for congruence with the second part. If, for some reason, it got rearranged, it would still be correct.  Happy to remove if it offends?  But that was my reasoning

thockin replied:
irrelevant now that the defer() is in.

thockin commented on pkg/registry/core/service/storage/alloc.go:
```diff
@@ -0,0 +1,1101 @@
+	}
+
+	// Allocate ports
+	if txn, err := al.txnAllocNodePorts(service, dryRun); err != nil {
```
This is a two-edged sword.  Then it becomes super easy to accidentally do:

```
var err error
var txn Transaction

// clean up on return
defer func() {
    if err != nil {
        result.Revert()
    }
}()

// correct
if txn, err = al.txnAllocClusterIPs(service, dryRun); err != nil {
    return nil, err
}
result = append(result, txn)

// broken!!
if txn, err := al.txnAllocNodePorts(service, dryRun); err != nil {
    return nil, err
}
result = append(result, txn)
```

The only difference is a sneaky `:`.

We could do something like:

```
var err error

// clean up on return
defer func() {
	if err != nil {
		result.Revert()
	}
}()
funcs := []func() (Transaction, error) {
	func() (Transaction, error) {
		return al.txnAllocClusterIPs(service, dryRun)
	},
	func() (Transaction, error) {
		return al.txnAllocNodePorts(service, dryRun)
	}
}

for _, fn := range funcs {
	var txn Transaction

	txn, err = fn()
	if err !=nil {
		return nil, err
	}
	result = append(result, txn)
}
```

but that feels equally convoluted.

What is your opinion?

thockin replied:
Or:

```
result := metaTransaction{}
success := false

defer func() {
    if !success {
        result.Revert()
    }
}()

if txn, err := al.txnAllocClusterIPs(service, dryRun); err != nil {
    return nil, err
}
result = append(result, txn)

// broken!!
if txn, err := al.txnAllocNodePorts(service, dryRun); err != nil {
    return nil, err
}
result = append(result, txn)

success = true
return result
```

That's not so bad.  I will add a new commit at the top of the stack.

thockin committed:
Svc REST: Make transaction-accumulating funcs safe

Identified in review, these funcs are now more reslient to future
changes.

thockin commented:
Rebased and pushed.  all comments addressed.  Great details on review.

thockin commented:
I have squashed FIXUP commits, updated all commit messages, and pushed - no code changes.

aojea commented:
/test pull-kubernetes-e2e-ipvs-azure-dualstack
/test pull-kubernetes-e2e-iptables-azure-dualstack
/test pull-kubernetes-e2e-kind-dual-canary
/test pull-kubernetes-e2e-kind-ipvs-dual-canary
/lgtm
/hold
holding waiting for CI and Kal
/honk

khenidak commented:
/lgtm
/approve
/unhold

here we go.. take cover!

khenidak commented:
/test pull-kubernetes-e2e-ubuntu-gce-network-policies
